File: apps/api/src/services/emailProviderResolver.ts
import {
  IntegrationModel,
  GmailProvider,
  OutlookProvider,
  SmtpProvider,
  decryptSecret,
  type EmailProvider,
} from "@believe-ai/server";
import { env } from "../config/env.js";

function buildGmailProvider(refreshToken: string): GmailProvider {
  if (!env.GMAIL_CLIENT_ID || !env.GMAIL_CLIENT_SECRET || !env.GMAIL_REDIRECT_URI) {
    throw new Error("Gmail OAuth is not configured on the server");
  }
  return new GmailProvider({
    clientId: env.GMAIL_CLIENT_ID,
    clientSecret: env.GMAIL_CLIENT_SECRET,
    redirectUri: env.GMAIL_REDIRECT_URI,
    refreshToken,
  });
}

function buildOutlookProvider(refreshToken: string): OutlookProvider {
  if (!env.OUTLOOK_CLIENT_ID || !env.OUTLOOK_CLIENT_SECRET || !env.OUTLOOK_REDIRECT_URI) {
    throw new Error("Outlook OAuth is not configured on the server");
  }
  return new OutlookProvider({
    clientId: env.OUTLOOK_CLIENT_ID,
    clientSecret: env.OUTLOOK_CLIENT_SECRET,
    redirectUri: env.OUTLOOK_REDIRECT_URI,
    refreshToken,
  });
}

/**
 * Same resolution logic as apps/worker/src/services/emailProviderResolver.ts
 * (campaign sends) — duplicated rather than shared because each app already
 * declares its own zod-validated env module, and the two call sites (a
 * queued campaign step vs. an immediate outreach send from a request) don't
 * share a runtime. Picks whichever provider the sending user actually
 * connected; never a hardcoded single provider.
 */
export async function resolveProvider(userId: string): Promise<EmailProvider> {
  if (env.EMAIL_PROVIDER === "smtp") {
    if (!env.SMTP_HOST || !env.SMTP_PORT) throw new Error("SMTP is not configured");
    return new SmtpProvider({ host: env.SMTP_HOST, port: env.SMTP_PORT, user: env.SMTP_USER, pass: env.SMTP_PASS });
  }

  const integration = await IntegrationModel.findOne({ userId, provider: { $in: ["gmail", "outlook"] } });
  if (!integration) {
    throw new Error("No email provider connected. Connect Gmail or Outlook in Settings before sending.");
  }

  const refreshToken = decryptSecret(integration.encryptedRefreshToken, env.ENCRYPTION_KEY);
  return integration.provider === "outlook" ? buildOutlookProvider(refreshToken) : buildGmailProvider(refreshToken);
}

File: apps/backend/services/email_provider_resolver.py
"""Picks whichever EmailProvider the sending user actually connected —
mirrors apps/worker's emailProviderResolver.ts (minus its EMAIL_PROVIDER=smtp
local-testing override, which isn't part of the production OAuth path)."""

from __future__ import annotations

from beanie.odm.operators.find.comparison import In
from bson import ObjectId

from core.config import Settings
from core.crypto import decrypt_secret
from core.errors import ValidationError
from email_providers.base import EmailProvider
from email_providers.gmail import GmailProvider
from email_providers.outlook import OutlookProvider
from models.integration import Integration


def _build_gmail_provider(settings: Settings, refresh_token: str) -> GmailProvider:
    if not (settings.gmail_client_id and settings.gmail_client_secret):
        raise ValidationError("Gmail OAuth is not configured on this server")
    return GmailProvider(settings.gmail_client_id, settings.gmail_client_secret, refresh_token)


def _build_outlook_provider(settings: Settings, refresh_token: str) -> OutlookProvider:
    if not (settings.outlook_client_id and settings.outlook_client_secret and settings.outlook_redirect_uri):
        raise ValidationError("Outlook OAuth is not configured on this server")
    return OutlookProvider(settings.outlook_client_id, settings.outlook_client_secret, settings.outlook_redirect_uri, refresh_token)


async def resolve_provider(settings: Settings, user_id: ObjectId) -> EmailProvider:
    integration = await Integration.find_one(Integration.userId == user_id, In(Integration.provider, ["gmail", "outlook"]))
    if not integration:
        raise ValidationError("No email provider connected. Connect Gmail or Outlook in Settings before sending.")

    assert settings.encryption_key is not None
    refresh_token = decrypt_secret(integration.encryptedRefreshToken, settings.encryption_key)
    if integration.provider == "outlook":
        return _build_outlook_provider(settings, refresh_token)
    return _build_gmail_provider(settings, refresh_token)
